### PR TITLE
Implementations of additional RNN architectures

### PR DESCRIPTION
**Pull request by transcranial**
Reference: An Empirical Exploration of Recurrent Network Architectures, Jozefowicz et al. 2015. http://www.jmlr.org/proceedings/papers/v37/jozefowicz15.pdf

They describe top RNN architectures evolved through the evaluation of thousands of models,  serving as alternatives to LSTMs and GRUs. These classes, `MutatedRNN_1`, `MutatedRNN_2`, `MutatedRNN_3` correspond to the `MUT1`, `MUT2`, and `MUT3` architectures described in the paper. Note that `input_dim` and `output_dim` must be equal in the `MUT1` and `MUT2` architectures.

I originally tried to implement these in a single class, but had to split them out.

In my tests on the IMDB dataset, everything being equal, these do improve on LSTM and GRU, particularly `MutatedRNN_1`.

### CODE REVIEW

File: docs/sources/layers/recurrent.md [hunk 1]
```diff
@@ -128,6 +128,39 @@ Long-Short Term Memory unit - Hochreiter 1997.
...
+## MutatedRNN_1, MutatedRNN_2, MutatedRNN_3
+
+```python
+keras.layers.recurrent.MutatedRNN_1(input_dim, output_dim=128, 
+        init='glorot_uniform', inner_init='orthogonal', 
+        activation='tanh', inner_activation='sigmoid',
+        weights=None, truncate_gradient=-1, return_sequences=False)
+```
+
+Top 3 RNN architectures evolved from the evaluation of thousands of models. Serves as alternatives to LSTMs and GRUs. Corresponds to `MUT1`, `MUT2`, and `MUT3` architectures described in the paper: An Empirical Exploration of Recurrent Network Architectures, Jozefowicz et al. 2015. Note that `input_dim` and `output_dim` must be equal in the `MUT1` and `MUT2` architectures.
```
Comment (fchollet) [inline docs/sources/layers/recurrent.md]: > Note that `input_dim` and `output_dim` must be equal in the `MUT1` and `MUT2` architectures.

Why? Highly impractical. And if that's the case, why are `input_dim` and `output_dim` two independent parameters in the constructor?

Comment (transcranial) [inline docs/sources/layers/recurrent.md]: In the equations some weight matrices are dropped. For example reset in `MUT2` is governed by `x_t + W_hr * h_tm1 + b_r`, as compared to `W_xr * x_t + W_hr * h_tm1 + b_r` in GRU. I suppose changing dimensionality would require another layer. This constraint may not be entirely in the spirit of keras, but they do perform quite well so I think it's worth including.

Right now an exception is raised if `input_dim` and `output_dim` are not equal, in the case of the first 2 architectures. Do you think it's better to have a single `dim` parameter?

Comment (fchollet) [inline docs/sources/layers/recurrent.md]: Couldn't we start with a matrix multiplication projecting the `input_dim`-dimensional input? onto an `output_dim`-dimensional space? It should not affect performance as long as the quantities involved are reasonable.

Comment (transcranial) [inline docs/sources/layers/recurrent.md]: I'll try that and see if it affects performance.

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -366,5 +366,306 @@ def get_config(self):
             "inner_activation":self.inner_activation.__name__,
             "truncate_gradient":self.truncate_gradient,
             "return_sequences":self.return_sequences}
+
+
+
+class MutatedRNN_1(Layer):
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Definitely not a fan of the names (in terms of style: it's a mixture of camel case and snake case, in terms of UX it's not descriptive, as the space of all possible genetically evolved RNNs is infinite).

Comment (fchollet) [inline keras/layers/recurrent.py]: We could give them nicknames I guess?

**Comment by fchollet** [comment]
> In my tests on the IMDB dataset, everything being equal, these do improve on LSTM and GRU, particularly MutatedRNN_1.

What other tasks have you tried them on? Why do you think the performance improvement is due to the architecture and not hyperparameter changes such as a higher hidden dimension?

**Comment by donglixp** [comment]
In the paper's conclusion section:
> We have evaluated a variety of recurrent neural network architectures in order to find an architecture that reliably outperforms the LSTM. Though there were architectures that outperformed the LSTM on some problems, we were unable to find an architecture that consistently beat the LSTM and the GRU in all experimental conditions.

Did you try the MUT1,MUT2,MUT3 models on your own tasks?

**Comment by transcranial** [comment]
Renamed the classes to JZS1, JZS2, JZS3, which combine the last names of the paper authors. Other suggestions welcome.

There's no more constraint of `input_dim` needing to be equal to `output_dim` with the first two architectures.

A head-to-head test run on the IMDB data against GRU with exactly the same hyperparameters give accuracies of:
GRU: 80.5
JZS1: 82.1
JZS2: 83.8
JZS3: 82.9

Some of my own tasks involve semantic classification of biomedical text, and on one sample problem, I get accuracies of: 
GRU: 90.1
JZS1: 90.2
JZS2: 90.8
JZS3: 91.1

Nothing mind-blowing, and probably very similar performance factoring in margin of error, but they are nonetheless cool to have as alternatives.

**Comment by fchollet** [comment]
I see GRU as an accuracy/speed tradeoff compared to LSTM; LSTM is clearly a superior architecture compared to GRU, however it is quite computationally intensive while GRU is much simpler and therefore much faster. So I am curious, do you have any benchmark against LSTM, and what is the time taken by these new layers compared to GRU and LSTM?

Also to note is that the IMDB dataset is too small to fully utilize LSTM... we don't have anything better in Keras though.

**Comment by transcranial** [comment]
Another run on the IMDB dataset:
LSTM: 81.4 (96s / epoch)
GRU: 83.4  (67s / epoch)
JZS1: 84.2 (58s / epoch)
JZS2: 82.9 (66s / epoch)
JZS3: 83.7 (69s / epoch)

So yeah, definitely within margin of error. JZS1 does offer speed advantages, though, even over GRU.

**Comment by fchollet** [comment]
That sounds good. I think we can merge this. 

By the way, did you seed the RNG for the IMDB experiments? The IMDB example script in Keras is not deterministic. 

I believe we should include the performance profile of recurrent layers in the documentation, on CPU and GPU, on a variety of tasks. That would be very useful to users.
